### DISCUSSION TITLE
extending PixiJS class to create multiple sprites with different textures

### DISCUSSION THREAD

**Discussion by ripmurdock** [pixijs/pixijs#9484] [Q&A] [answered]
I'd like to revise this PixiJS example codepen to use multiple textures.
https://codepen.io/ripmurdock/pen/wvQGjoW?editors=0011

The AwesomeSprite class extends PIXI.Sprite.  I expected that each sprite created with this class would use the texture that is passed at the time the class is called.

When the class is called multiple times to create multiple instances with different texture parameters, every instance uses the first texture passed.

I've tried creating multiple classes with unique class names.  Even if the classes are duplicated, PIXI.Sprite creates a new sprite with the first texture that is passed.

How can I revise this code, so that PIXI.Sprite creates a new sprite with the current texture parameter each time the AwesomeSprite class is called?

Thanks!

Here's the javascript from codepen:

gsap.config({trialWarn: false})
console.clear();

asyncCall();

async function asyncCall() {
let minX = 0;
let minY = 0;
let maxX = window.innerWidth;
let maxY = window.innerHeight;

let total   = 0;
let batch   = 3;
let limit   = 5000;

let ease  = Power0.easeNone;  

const app = new PIXI.Application({autoResize: true, backgroundAlpha: 0});

app.renderer.resize(maxX, maxY);

PIXI.Assets.addBundle('texturesBundle', {
    bunny: 'https://pixijs.io/examples/examples/assets/bunny.png',
    eggHead: 'https://pixijs.io/examples/examples/assets/eggHead.png',
    flowerTop: 'https://pixijs.io/examples/examples/assets/flowerTop.png', 
  });

const assets = await PIXI.Assets.loadBundle('texturesBundle');

let sprites = [];

document.body.appendChild(app.view);  

let container = new PIXI.ParticleContainer(limit, {
  scale: true,
  position: true,
  rotation: true,
  uvs: false,
  alpha: true
});

app.stage.addChild(container);

//
// AwesomeSprite
// ========================================================================
class AwesomeSprite extends PIXI.Sprite {
  constructor(texture) {
    super(texture);       
    let repeat = -1;
    let yoyo = true;
    let time1  = Math.random()*(3-0.5)+0.5
    let time2  = Math.random()*(4.5-2)+2
    let time3  = Math.random()*(4.5-2)+2
    let prog1  = Math.random()
    let prog2  = Math.random()
    let prog3  = Math.random()
    let scale  = Math.random()*0.5
    let alpha  = 1
    let speedX = (Math.random()*(10+10)-10) * (-1 + Math.round(Math.random()) * 2) 
    let speedY = (Math.random()*(10-2)+2) * (-1 + Math.round(Math.random()) * 2)   
    let startX = .5*(maxX)
    let startY = .5*(maxY)
    let angle  = (-1 + Math.round(Math.random()) * 2) * Math.PI * 2

    this.anchor.set(0.5);
    this.pivot.set(0.5);
    this.scale.set(0.1);
    
    this.x = startX;
    this.y = startY; 
    this.alpha = alpha;               
    this.rotation = 0;    
    
    this.speed = new PIXI.Point(speedX, speedY);
        
    gsap.to(this,{ repeat, ease, rotation: angle, duration: time1, }).progress(prog1);    
    gsap.to(this.scale,{ repeat, ease, yoyo, x: 1, y: 1, duration: time3, }).progress(prog3);   
  }
  
  update() {
    
    this.x += this.speed.x;
    this.y += this.speed.y;

    if (this.x > maxX) {
      
      this.x = maxX;
      this.speed.x *= -1;
      
    } else if (this.x < minX) {
      
      this.x = minX;
      this.speed.x *= -1;
    }
    
    if (this.y > maxY) {
      
      this.y = maxY;
      this.speed.y *= -1;
      
    } else if (this.y < minY) {
      
      this.y = minY;
      //this.speed.y = 0;
      //this.speed.y = 0;
      this.speed.y *= -1;
    }
  }
}

  gsap.ticker.add(render);
    createSprites(batch);
    let sprite = PIXI.Sprite.from(assets.bunny);
    sprite.anchor.set(0.5);
    sprite.x = app.screen.width * 2/9;
    sprite.y = app.screen.height / 2;
    app.stage.addChild(sprite);
  
    sprite = PIXI.Sprite.from(assets.eggHead);
    sprite.anchor.set(0.5);
    sprite.x = app.screen.width * 4/9;
    sprite.y = app.screen.height / 2;
    app.stage.addChild(sprite);
  
    sprite = PIXI.Sprite.from(assets.flowerTop);
    sprite.anchor.set(0.5);
    sprite.x = app.screen.width * 6/9;
    sprite.y = app.screen.height / 2;
    app.stage.addChild(sprite);
  
//
// CREATE sprites
// ========================================================================
function createSprites(count) {
  //console.log(count)
  for (let i = 0; i < count; i++) {    
   createSprite(i)
  } 
  
  function createSprite(i) {
    let sprite0
    switch(i%3) {
      case 0:
        sprite0 = new AwesomeSprite(assets.eggHead);
        break;
      case 1:
        sprite0 = new AwesomeSprite(assets.bunny);
        break;
      default:
        sprite0 = new AwesomeSprite(assets.flowerTop);
    }
        
    sprites.push(sprite0);
    container.addChild(sprite0);
    total++
  }
}

//
// RENDER
// ========================================================================
function render() {

  for (let i = 0; i < total; i++) {
    sprites[i].update();
  }
  
  app.renderer.render(app.stage);
}
}

**Comment by andrewstart** [comment, marked as answer]
`PIXI.ParticleContainer` is probably the issue - IIRC, it only allows a single texture among all of its children. You should be able to use multiple textures in an atlas, but not multiple separate images.

**Reply by ripmurdock** [reply]
Thanks!  It turns out it was the one texture per ParticleContainer rule that was the culprit.

I've gotten this animation to work by creating a ParticleContainer for each texture:
https://codepen.io/ripmurdock/pen/jOQqRQa?editors=0011

I'd like to understand how to use the @pixi/layers module to reorder the sprites based on their current scale.

This line appears to successfully add the Layer function in codepen:
import { Layer, applyParticleMixin   }  from "https://esm.sh/@pixi/layers@2.0.1"

I'm guessing "applyCanvasMixin" here is a typo in the docs:

import { ParticleContainer } from 'pixi.js';
import { applyCanvasMixin } from '@pixi/layers';
applyParticleMixin(ParticleContainer);

If I add this line,
applyParticleMixin(PIXI.ParticleContainer);

The log generates this error:
Uncaught TypeError: r.incDisplayOrder is not a function

This animation is already pretty close to what I'm going for.  If I can order the sprites by scale, that will take it to the next level.

Thanks again for your help!

gsap.config({trialWarn: false})
console.clear();

import { Layer, applyParticleMixin   }  from "https://esm.sh/@pixi/layers@2.0.1"
//applyParticleMixin(PIXI.ParticleContainer);

asyncCall();

async function asyncCall() {
let minX = 0;
let minY = 0;
let maxX = window.innerWidth;
let maxY = window.innerHeight;

let total   = 0;
let batch   = 30;
let limit   = 5000;

let ease  = Power0.easeNone;  

const app = new PIXI.Application({autoResize: true, backgroundAlpha: 0});

app.renderer.resize(maxX, maxY);

PIXI.Assets.addBundle('texturesBundle', {
    bunny: 'https://pixijs.io/examples/examples/assets/bunny.png',
    eggHead: 'https://pixijs.io/examples/examples/assets/eggHead.png',
    flowerTop: 'https://pixijs.io/examples/examples/assets/flowerTop.png', 
  });

const assets = await PIXI.Assets.loadBundle('texturesBundle');

let sprites = [], textures = [], particleContainers = [];

//
// AwesomeSprite
// ========================================================================
class AwesomeSprite extends PIXI.Sprite {
  constructor(texture) {
    super(texture);       
    let repeat = -1;
    let yoyo = true;
    let time1  = Math.random()*(3-0.5)+0.5
    let time2  = Math.random()*(4.5-2)+2
    let time3  = Math.random()*(4.5-2)+2
    let prog1  = Math.random()
    let prog2  = Math.random()
    let prog3  = Math.random()
    let scale  = Math.random()*0.5
    let alpha  = 1
    //let speedX = (Math.random()*(10+10)-10) * (-1 + Math.round(Math.random()) * 2) 
    //let speedY = (Math.random()*(10-2)+2) * (-1 + Math.round(Math.random()) * 2)  
    let speedX = (Math.random()*(1+1)-1) * (-1 + Math.round(Math.random()) * 2) 
    let speedY = (Math.random()*(1-.1)+.1) * (-1 + Math.round(Math.random()) * 2) 
    let startX = .5*(maxX)
    let startY = .5*(maxY)
    let angle  = (-1 + Math.round(Math.random()) * 2) * Math.PI * 2
    
    this.zOrder = scale;
    this.anchor.set(0.5);
    this.pivot.set(0.5);
    this.scale.set(0.1);
    
    this.x = startX;
    this.y = startY; 
    this.alpha = alpha;               
    this.rotation = 0;    
    
    this.speed = new PIXI.Point(speedX, speedY);
        
    gsap.to(this,{ repeat, ease, rotation: angle, duration: time1, }).progress(prog1);    
    gsap.to(this.scale,{ repeat, ease, yoyo, x: 1, y: 1, duration: time3, }).progress(prog3);   
  }
  
  update() {
    this.zOrder = this.scale;
    this.x += this.speed.x;
    this.y += this.speed.y;

    if (this.x > maxX) {
      
      this.x = maxX;
      this.speed.x *= -1;
      
    } else if (this.x < minX) {
      
      this.x = minX;
      this.speed.x *= -1;
    }
    
    if (this.y > maxY) {
      
      this.y = maxY;
      this.speed.y *= -1;
      
    } else if (this.y < minY) {
      
      this.y = minY;
      //this.speed.y = 0;
      //this.speed.y = 0;
      this.speed.y *= -1;
    }
  }
}

    textures[0] = (assets.eggHead);
    textures[1] = (assets.flowerTop);
    textures[2] = (assets.bunny);
    const layer = new Layer();
    layer.group.enableSort = true;
    for (let j=0;j<textures.length;j++){
    particleContainers[j] = new PIXI.ParticleContainer(limit, {
              scale: true,
              position: true,
              rotation: true,
              uvs: false,
              alpha: true
              });
    app.stage.addChild(particleContainers[j]);
    createSprites(batch, textures[j], particleContainers[j], layer);
    }
  
    document.body.appendChild(app.view);  
    gsap.ticker.add(render);
//
// CREATE sprites
// ========================================================================
function createSprites(count, texture, container, layer) {
  //console.log(count)
  for (let i = 0; i < count; i++) {    
   createSprite(texture, container, layer)
  } 
  function createSprite(texture, container, layer) {
    const sprite = new AwesomeSprite(texture);
    sprite.parentLayer = layer;
    sprites.push(sprite);
    container.addChild(sprite);
    total++
  }
}

//
// RENDER
// ========================================================================
function render() {

  for (let i = 0; i < total; i++) {
    sprites[i].update();
  }
  
  app.renderer.render(app.stage);
  console.log(layer._sortedChildren);
}
}
